fabric-client-gametest-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/client/gametest/ClientGameTestTest.java

```diff
@@ -16,141 +16,85 @@
+			{
+				enableDebugHud(context);
+				singleplayer.getClientWorld().waitForChunksRender();
+				context.takeScreenshot("in_game_overworld", 0);
+			}
 
-			final GameProfile profile = context.computeOnClient(MinecraftClient::getGameProfile);
-			server.runCommand("op " + profile.getName());
-			server.runCommand("gamemode creative " + profile.getName());
+			{
+				context.getInput().pressKey(options -> options.chatKey);
+				context.waitTick();
+				context.getInput().typeChars("Hello, World!");
+				context.getInput().pressKey(InputUtil.GLFW_KEY_ENTER);
+				context.takeScreenshot("chat_message_sent", 5);
+			}
 
-			waitForWorldTicks(context, 20);
-			context.takeScreenshot("server_in_game", 0);
+			MixinEnvironment.getCurrentEnvironment().audit();
 
-			{ // Test that we can enter and exit configuration
-				server.runCommand("debugconfig config " + profile.getName());
-				context.waitForScreen(ReconfiguringScreen.class);
-				context.takeScreenshot("server_config");
-				server.runCommand("debugconfig unconfig " + profile.getId());
-				waitForWorldTicks(context, 1);
+			{
+				// See if the player render events are working.
+				setPerspective(context, Perspective.THIRD_PERSON_BACK);
+				context.takeScreenshot("in_game_overworld_third_person");
+				setPerspective(context, Perspective.FIRST_PERSON);
 			}
 
-			context.setScreen(() -> new GameMenuScreen(true));
-			context.takeScreenshot("server_game_menu");
-			context.clickScreenButton("menu.disconnect");
-
-			context.waitForScreen(MultiplayerScreen.class);
-			context.clickScreenButton("gui.back");
+			{
+				context.getInput().pressKey(options -> options.inventoryKey);
+				context.takeScreenshot("in_game_inventory");
+				context.setScreen(() -> null);
+			}
 		}
 
-		{
-			context.waitForScreen(TitleScreen.class);
+		try (TestSingleplayerContext singleplayer = spWorldSave.open()) {
+			singleplayer.getClientWorld().waitForChunksRender();
+			context.takeScreenshot("in_game_overworld_2");
 		}
-	}
 
-	private static boolean isDirEmpty(Path path) {
-		try (DirectoryStream<Path> directory = Files.newDirectoryStream(path)) {
-			return !directory.iterator().hasNext();
-		} catch (IOException e) {
-			throw new UncheckedIOException(e);
+		try (TestDedicatedServerContext server = context.worldBuilder().createServer()) {
+			try (TestServerConnection connection = server.connect()) {
+				connection.getClientWorld().waitForChunksRender();
+				context.takeScreenshot("server_in_game", 0);
+
+				{ // Test that we can enter and exit configuration
+					final GameProfile profile = context.computeOnClient(MinecraftClient::getGameProfile);
+					server.runCommand("debugconfig config " + profile.getName());
+					context.waitForScreen(ReconfiguringScreen.class);
+					context.takeScreenshot("server_config");
+					server.runCommand("debugconfig unconfig " + profile.getId());
+					// TODO: better way to wait for reconfiguration to end
+					context.waitTicks(100);
```

Comment:
Not having this `waitTicks(100)` here breaks the assertion in `TestServerConnection.close()` because reconfiguration sets `MinecraftClient.world` to `null`. Ideally I want the `close()` method to wait for any configuration to finish before trying to disconnect. I tried to fix this for this PR, but I didn't manage to figure out a good way to fix it, so I've left the problem for later. I imagine it's a pretty rare case anyway.

Reply:
Yeah, could you maybe wait until the world has rendered? I dont think it matters a vast amount, as this test is something that I dont expect many if anyone to run. I think I added it after we broke reconfiguration, im not even sure the debugconfig command is registered by default.